### PR TITLE
[dynamo] GradModeVariable should only be eagerly initialized when doing the equivalent of `set_grad_enabled`

### PR DESCRIPTION
**Pull request by jon-chuang**
Grad mode variable was previously initialized eagerly when called - which is wrong when not explicitly using it in `set_grad_enabled`

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_dynamo/variables/ctx_manager.py [hunk 1]
```diff
@@ -181,6 +181,14 @@ def exit(self, tx, *args):
         self._call_func(tx, self.initial_values)
         return variables.ConstantVariable.create(None)
 
+    def call_function(
+        self, tx, args: "List[VariableTracker]", kwargs: "Dict[str, VariableTracker]"
+    ):
+        # TODO(jon-chuang): uncomment once https://github.com/pytorch/pytorch/issues/113298 is fixed
+        # self._call_func(tx, self.initial_values)  # undo eager initialization
+        self.initialized = False
```
Comment (jansel) [inline torch/_dynamo/variables/ctx_manager.py:189]: Mutating variable trackers in unsafe because we make copies of them frequently.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_dynamo/variables/ctx_manager.py (hunk 1 above)
Comment (jon-chuang) [reply torch/_dynamo/variables/ctx_manager.py:189]: Actually, for now, I think we don't need to mutate anything. We may need to if and when https://github.com/pytorch/pytorch/issues/113298 is fixed.

Comment (jon-chuang) [reply torch/_dynamo/variables/ctx_manager.py]: removed the initialized var in favour of statelessly always running `_call_func`

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**